openconcept/aerodynamics/openaerostruct/mesh_gen.py

```diff
@@ -0,0 +1,707 @@
+
+        # Compute derivatives in a way analogous to forward AD
+        db_dS = AR / (4 * np.sqrt(AR * S))
+        db_dAR = S / (4 * np.sqrt(AR * S))
+        dcroot_dS = 1 / (half_span * (1 + taper)) - S / (half_span**2 * (1 + taper)) * db_dS
+        dcroot_dAR = -S / (half_span**2 * (1 + taper)) * db_dAR
+        dcroot_dtaper = -S / (half_span * (1 + taper) ** 2)
+
+        dy_dS = y_mesh * db_dS
+        dy_dAR = y_mesh * db_dAR
+
+        dx_dS = x_mesh * np.linspace(taper, 1, self.ny).reshape(1, self.ny) * dcroot_dS
+        dx_dS -= np.linspace(dcroot_dS * taper, dcroot_dS, self.ny).reshape(1, self.ny) / 4
+        dx_dS += np.linspace(db_dS, 0, self.ny).reshape(1, self.ny) * np.tan(np.deg2rad(sweep))
+
+        dx_dAR = x_mesh * np.linspace(taper, 1, self.ny).reshape(1, self.ny) * dcroot_dAR
+        dx_dAR -= np.linspace(dcroot_dAR * taper, dcroot_dAR, self.ny).reshape(1, self.ny) / 4
+        dx_dAR += np.linspace(db_dAR, 0, self.ny).reshape(1, self.ny) * np.tan(np.deg2rad(sweep))
+
+        dx_dtaper = (
+            x_mesh * c_root * np.linspace(1, 0, self.ny).reshape(1, self.ny)
+            + x_mesh * np.linspace(taper, 1, self.ny).reshape(1, self.ny) * dcroot_dtaper
+        )
+        dx_dtaper -= (
+            np.linspace(c_root, 0, self.ny).reshape(1, self.ny) / 4
+            + np.linspace(dcroot_dtaper * taper, dcroot_dtaper, self.ny).reshape(1, self.ny) / 4
+        )
+
+        dx_dsweep = (
+            0 * x_mesh
+            + np.linspace(half_span, 0, self.ny).reshape(1, self.ny) / np.cos(np.deg2rad(sweep)) ** 2 * np.pi / 180.0
+        )
+
+        J["mesh", "S"] = np.dstack((dx_dS, dy_dS, np.zeros((self.nx, self.ny)))).flatten()
+        J["mesh", "AR"] = np.dstack((dx_dAR, dy_dAR, np.zeros((self.nx, self.ny)))).flatten()
+        J["mesh", "taper"] = np.dstack(
+            (dx_dtaper, np.zeros((self.nx, self.ny)), np.zeros((self.nx, self.ny)))
+        ).flatten()
+        J["mesh", "sweep"] = np.dstack(
+            (dx_dsweep, np.zeros((self.nx, self.ny)), np.zeros((self.nx, self.ny)))
+        ).flatten()
+
+
+class SectionPlanformMesh(om.ExplicitComponent):
+    """
+    Generate an OpenAeroStruct mesh based on defined section parameters, similar
+    to how AVL accepts planforms. The resulting mesh is for a half wing (meant
+    to use with OpenAeroStruct symmetry), but the input reference area is for the
+    full wing.
+
+    The sectional properties (x offset, y value, and chord) are nondimensional,
+    but their relative magnitudes reflect the actual shape. The full planform
+    is scaled to match the requested planform area. Optionally, this scaling
+    can be turned off (by setting scale_area=False) and the planform area is
+    then computed as an output.
+
+    The mesh points are cosine spaced in the chordwise direction and in the
+    spanwise direction within each trapezoidal region of the wing defined by
```

Comment:
Could the spacing be a parameter for the user to decide? Cosine spacing might lead to weird distributions for a two-section wing.

Reply:
AVL actually uses a per-section cosine distribution, which is why I took this approach. It becomes cumbersome to make it super general which is why for now I stuck to only cosine spaced. Cosine will give a better result that linear spacing.

Reply:
See "Vortex Lattice Spacing Distributions" here: https://web.mit.edu/drela/Public/web/avl/avl_doc.txt

Reply:
Okay, as long as this is normal convention and leads to good results then sounds good to me!